cupy/cuda/compiler.py

```diff
@@ -274,6 +276,102 @@ def _hash_hexdigest(value):
 
 
 _hash_length = len(_hash_hexdigest(b''))  # 40 for SHA1
+_pickle_protocol = pickle.HIGHEST_PROTOCOL  # depends on actual python version
+
+
+class RestrictedUnpickler(pickle.Unpickler):
+    """An unpickler that only allow some safe classes."""
+
+    def find_class(self, module, name):
+        """Only allow a restricted set of classes."""
+        if module == __name__ and name == 'CachedModule':
+            return CachedModule
+        msg = "'{}.{}' is forbidden and cannot be unpickled."
+        raise pickle.UnpicklingError(msg.format(module, name))
+
+    @classmethod
+    def loads(cls, s):
+        """Helper function analogous to pickle.loads()."""
+        return cls(io.BytesIO(s)).load()
+
+
+class CachedModule:
+    __slots__ = ['backend', 'binary', 'mapping']
+
+    def __init__(self, backend, binary, mapping):
+        self.backend = backend
+        self.binary = binary
+        self.mapping = mapping
+
+    @classmethod
+    def dump_module(cls, path, backend, binary, mapping):
+        """Build and dumps a CachedModule object and checksum to path."""
+        python_obj = cls(backend, binary, mapping)
+
+        data_bytes = pickle.dumps(python_obj, protocol=_pickle_protocol)
+        data_hash = _hash_hexdigest(data_bytes).encode('ascii')
+        assert len(data_hash) == _hash_length
+
+        directory = os.path.dirname(path)
+        if not os.path.isdir(directory):
+            os.makedirs(directory, exist_ok=True)
+
+        # We use a temp file to handle concurrent processes caching
+        # to the same path at the same time.
+        with tempfile.NamedTemporaryFile(
+                dir=directory, delete=False, mode='wb') as tf:
+            tf.write(data_hash)
+            tf.write(data_bytes)
+            temp_path = tf.name
+
+        # os.replace is not atomic operation, so it could result in a
```

Comment:
`os.replace` is an atomic operation, so checksum in cache file is no longer needed 😄  (see #4926)

Reply:
Yup when I left the comment to @keckj I forgot about this. @kmaehashi, do you suggest to revise the comment here, or to remove the hash prefix entirely?

Reply:
I think it's ok to entirerly remove hash prefix.

Reply:
I removed the hash. What should happen if mod.load(cached_module.binary) fails ?

Reply:
Thanks! I think it's safe to assume that the file is not broken.